elf_load.c

```diff
@@ -210,110 +196,588 @@ struct elf_prog *elf_load_img(struct uk_alloc *a, void *img_base,
+		goto err_free_elf_prog;
+	}
+	if (elf_prog->interp.required) {
+		uk_pr_err("%s: Requests program interpreter: Unsupported for in-memory ELF images\n",
+			  progname);
+		ret = -ENOTSUP;
+		goto err_free_elf_prog;
 	}
 
-out_free_elf:
+	ret = elf_load_imgcpy(elf_prog, elf, img_base, img_len);
+	if (ret < 0) {
+		uk_pr_err("%s: Failed to copy the executable: %d\n",
+			  progname, ret);
+		goto err_free_elf_prog;
+	}
+
+	ret = elf_load_ptprotect(elf_prog, elf);
+	if (ret < 0) {
+		uk_pr_err("%s: Failed to set page protection bits: %d\n",
+			  progname, ret);
+		goto err_unload_vaimg;
+	}
+
+	elf_end(elf);
+	return elf_prog;
+
+err_unload_vaimg:
+	elf_unload_vaimg(elf_prog);
+err_free_elf_prog:
+	uk_free(a, elf_prog);
+err_end_elf:
+	elf_end(elf);
+err_out:
+	return ERR2PTR(ret);
+}
+
+#if CONFIG_LIBVFSCORE
+static struct elf_prog *do_elf_load_vfs(struct uk_alloc *a, const char *path,
+					const char *progname, bool nointerp)
+{
+	int fd = -1;
+#if CONFIG_APPELFLOADER_VFSEXEC_EXECBIT
+	struct stat fd_stat;
+#endif /* CONFIG_APPELFLOADER_VFSEXEC_EXECBIT */
+	struct elf_prog *elf_prog = NULL;
+	Elf *elf;
+	int ret;
+
+	fd = open(path, O_RDONLY);
+	if (fd < 0) {
+		uk_pr_err("%s: Failed to execute %s: %s\n",
+			  progname, path, strerror(errno));
+		ret = -errno;
+		goto err_out;
+	}
+
+#if CONFIG_APPELFLOADER_VFSEXEC_EXECBIT
+	/* Check for executable bit */
+	ret = fstat(fd, &fd_stat);
+	if (ret != 0) {
+		uk_pr_err("%s: Failed to execute %s: %s\n",
+			  progname, path, strerror(errno));
+		ret = -errno;
+		goto err_close_fd;
+	}
+	if (!(fd_stat.st_mode & S_IXUSR)) {
+		uk_pr_err("%s: Failed to execute %s: %s\n",
+			  progname, path, strerror(EPERM));
+		ret = -EPERM;
+		goto err_close_fd;
+	}
+#else /* !CONFIG_APPELFLOADER_VFSEXEC_EXECBIT */
+	uk_pr_debug("%s: Note, ignoring executable bit state\n", progname);
+#endif /* !CONFIG_APPELFLOADER_VFSEXEC_EXECBIT */
+
+	elf = elf_open(fd);
+	if (!elf) {
+		elferr_err("%s: Failed to initialize ELF parser\n",
+			   progname);
+		ret = -EBUSY;
+		goto err_close_fd;
+	}
+
+	elf_prog = uk_calloc(a, 1, sizeof(*elf_prog));
+	if (!elf_prog) {
+		ret = -ENOMEM;
+		goto err_end_elf;
+	}
+	elf_prog->a = a;
+	elf_prog->name = progname;
+
+	ret = elf_load_parse(elf_prog, elf);
+	if (ret < 0) {
+		uk_pr_err("%s: Parsing of ELF image failed: %s (%d)\n",
+			  progname, strerror(-ret), ret);
+		goto err_free_elf_prog;
+	}
+	if (nointerp && elf_prog->interp.required) {
+		uk_pr_err("%s: Requests program interpreter: Unsupported\n",
+			  progname);
+		ret = -ENOTSUP;
+		goto err_free_elf_prog;
+	}
+
+	ret = elf_load_fdread(elf_prog, elf, fd);
+	if (ret < 0) {
+		uk_pr_err("%s: Failed to copy the executable: %d\n",
+			  progname, ret);
+		goto err_free_elf_prog;
+	}
+
+	ret = elf_load_ptprotect(elf_prog, elf);
+	if (ret < 0) {
+		uk_pr_err("%s: Failed to set page protection bits: %d\n",
+			  progname, ret);
+		goto err_unload_vaimg;
+	}
+
+	elf_end(elf);
+	close(fd);
+	elf_prog->path = path;
```

Comment:
Is there a reason why this only happens at this point and not already in L702?

Reply:
Ah, good point. Reason was originally the error clean-up code. But I can move it up, looks cleaner then.